### DISCUSSION TITLE
Large-data survival forests: how to tune parameters efficiently?

### DISCUSSION THREAD

**Discussion by simontrlifaj** [kogalur/randomForestSRC#222] [Q&A] [answered]
Hi, thanks for the excellent package and documentation, its very useful! I have a few questions about working with large-data survival forests.

I am training survival random forests on N = 4,101,858 patent data (context: the survival event is the ownership change of a patent). I have 11 variables + two factors with around 35 levels each. My main goal is to find a combination of parameters that results in high Harrell's c-index and then use the resulting forest to analyse variable importance. I want to reduce compute time to be able to quickly iterate between different parameters, but I keep running into problems with larger sample sizes.

Setup: I am using the `rfsrc.fast` function with `samptype = "swor"`, `perf.type="none"`, and `importance = "none"`. I run it on a train subsample of my data of size 10,000 to 500,000 (see question 1. below). My default parameters are `ntree = 1000`, `nodesize = 200`, `nodedepth = 6`, `ntime = 150`, `nsplit = 30` and `sampsize = 1000`, with which I try to experiment.

Questions:
1. Whenever I pass a larger (> 200,000) dataset to `predict.rfsrc`, the compute time dramatically increases, even with no changes to `sampsize` and other parameters. This prevents me from using my whole dataset. Why could that be when I am not computing c-index and variable importance? I thought there would be no effect, as each tree trains only on a subset given by `sampsize`.
2. What is the most efficient way to calculate the c-index (to evaluate the model outcome), without calculating the variable importance? My current setup: Following instructions [here](https://www.randomforestsrc.org/articles/speedup.html), I turned `perf.type="none"` in `rfsrc.fast`. I then run `predict.rfsrc`  and set `sampsize` to something like 1000. I also run it only on a test subsample (as in 1., running it on the full data results in high compute time). Is there a more efficient way to quickly evaluate a forest?
3. Could setting `ntime` to lower values result in higher c-index? Setting it low reduces compute time, but I don't want to be getting unreliable c-indexes. In the data, I have 7,300 distinct time points.
4. I kept having issues with the factor variables, so I turned them into one hot encoding. Is that equivalent to passing them as factors?

Any advice is much appreciated!

**Comment by ishwaran** [comment, marked as answer]
Prediction can be intensive with a large prediction data set because there's alot going on under the hood.  However, there is a hidden option in prediction that might help.  In your prediction call try `predict(object, newdata, jitt = FALSE)` and see if that improves things.

Hot encoding (i.e. ANOVA dummy encoding) is a terrible way to deal with factors and I would never recommend it if your goal is to assess the importance of a factor.  It seems like you are splitting the original data into train/test datasets and what might be happening is that the test data has factor levels not seen in the training data.  This will throw up an error `"levels of factors in test data do not match those in training data"`.  If this is what you are seeing, then one solution (if it is possible due to memory) is to store your data set as a data frame in R and then when you split the data in train/test groups the factor levels will be correctly encoded across both data sets and the error should no longer happen.

Your main issue is that you want a fast way to test many different models and assess which one is best.  But the problem is that the C-index is slow.  Therefore the solution is not to use the C-index.  Use the same settings you mention above to grow the forest, in particular `perf.type="none"` (note that importance is always off by default).  Then use `get.brier.survival(o)` on your grow forest object `o`  to obtain the brier score and CRPS which are excellent ways of measuring performance.  By default `get.brier.survival`  uses the OOB survival so the performance values will be out of sample.   See the [random survival forests vignette](https://www.randomforestsrc.org/articles/survival.html) for more details about the Brier score.

**Reply by simontrlifaj** [reply]
Thanks a lot, very useful!

**Reply by ishwaran** [reply]
I want to mention that the latest release of the software can handle levels of factors in test data not matching training data.

https://www.randomforestsrc.org/articles/getstarted.html#allowable-data-types-and-factors

**Reply by simontrlifaj** [reply]
The issue with factors had a different solution: the data were in a grouped_df format (because of some previous dplyr manipulations). All I needed to do is to convert to standard data.frame. (Before, it did not give any error, just crushed.) But the new feature seems definitely useful, especially when one works on the code on a small subsample, the likelihood of mismatch between train and test set is high.

**Reply by ishwaran** [reply]
Right we've had this comment on the discussion board about having to convert to data frame when using dplyr.

Regarding the new robust treatment of factors.  In my experience it happens more often than people realize.  This is because you often find data coding errors when factors are input.  So for example "cat" could be input "Cat" and now you have a problem.

**Reply by simontrlifaj** [reply]
One more thing about using `brier` to evaluate trees: is it possible to use brier score to calculate variable importance (with confidence intervals)? I tried passing it to `subsample`, but that did not accept it. `holdout.vimp` does accept it under `perf.type`, but it does not calculate confidence intervals.

**Comment by ishwaran** [comment]
`subsample` does allow `perf.type`

```
> o <- rfsrc(Species~.,iris,importance="permute",perf.type="brier")
> so <- subsample(o)
> plot.subsample(so, standardize=FALSE)

```

You can compare the results with standard misclassification error by replacing the grow object above with

`> o <- rfsrc(Species~.,iris,importance="permute")`

**Reply by simontrlifaj** [reply]
Thanks, but I can't make that work for survival forests.

Example: adding `perf.type="brier"` and `importance = "permutate"` to the example from the [survival vignette](https://www.randomforestsrc.org/articles/survival.html):
```
data(peakVO2, package = "randomForestSRC")
dta <- peakVO2
obj <- randomForestSRC::rfsrc(Surv(ttodead,died)~., dta,
             ntree = 1000, nodesize = 5, nsplit = 50, importance = "permutate", perf.type="brier")
```
Returns `Error in match.arg(perf, c("none", "default", "standard")) : 'arg' should be one of “none”, “default”, “standard”`.

**Comment by ishwaran** [comment]
That's because `perf.type="brier"` applies to classification only.  Sorry, I thought you were using classification analysis - as currently there is no brier survival splitting rule.  Also it's `permute` and not permutate

**Reply by simontrlifaj** [reply]
Ok, thanks for the clarification. I am now able to train the forests on ~ 500 k sample, and calculate importance with `vimp`. Still, getting the confidence intervals with `subsample` just never finishes calculating importance. So there is no other way to get confidence intervals, to get them based on a subsample of the training trees / data, or to base them on another metric than concordance?

**Comment by ishwaran** [comment]
I've been thinking about this, but let me get a few things clear here.  You say you are able to calculate variable importance on your ~500K sample using `vimp`?  Because if you are able to do that then `subsample` should be quite rapid.  The one issue with subsample is it needs one call to the original data to get variable importance from the full data set.  This full data estimator is used for the pivot value when calculating the confidence region (subsampling is used to obtain the standard error). So typically you would make your original training call with `importance=TRUE` and that way there is no other time that the full data is needed.  So all the remaining calculations are on subsampled data where presumably calculations will be rapid.

Here's an illustration of how this works for the PBC data

```
> data(pbc, package = "randomForestSRC.stealth")
> srv.o <- rfsrc(Surv(days, status) ~ ., pbc, importance = TRUE)
> srv.smp.o <- subsample(srv.o, B = 100, verbose = TRUE)

```
